layer_analyzer_test.go

```diff
@@ -0,0 +1,303 @@
+					// Cache-image-only layers.
+					{"metadata.buildpack/cache.sha", "cache-sha"},
+				} {
+					got := h.MustReadFile(t, filepath.Join(layerDir, data.name))
+					h.AssertStringContains(t, string(got), data.want)
+				}
+			})
+
+			it("does not overwrite metadata from app image", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				for _, name := range []string{
+					"metadata.buildpack/launch-build-cache.toml",
+					"metadata.buildpack/launch-cache.toml",
+				} {
+					got := h.MustReadFile(t, filepath.Join(layerDir, name))
+					avoid := "[metadata]\n  cache-only-key = \"cache-only-value\""
+					if strings.Contains(string(got), avoid) {
+						t.Errorf("Expected %q to not contain %q, got %q", name, avoid, got)
+					}
+				}
+			})
+
+			it("does not overwrite sha from app image", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				for _, name := range []string{
+					"metadata.buildpack/launch-build-cache.sha",
+					"metadata.buildpack/launch-cache.sha",
+				} {
+					got := h.MustReadFile(t, filepath.Join(layerDir, name))
+					avoid := "old-sha"
+					if strings.Contains(string(got), avoid) {
+						t.Errorf("Expected %q to not contain %q, got %q", name, avoid, got)
+					}
+				}
+			})
+
+			it("does not restore cache=true layers for non-selected groups", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				h.AssertPathDoesNotExist(t, filepath.Join(layerDir, "no.group.buildpack"))
+			})
+
+			it("does not restore launch=true layer metadata", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				h.AssertPathDoesNotExist(t, filepath.Join(layerDir, "metadata.buildpack", "launch-cache-not-in-app.toml"))
+				h.AssertPathDoesNotExist(t, filepath.Join(layerDir, "metadata.buildpack", "launch-cache-not-in-app.sha"))
+			})
+
+			it("does not restore cache=false layer metadata", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				h.AssertPathDoesNotExist(t, filepath.Join(layerDir, "metadata.buildpack", "cache-false.toml"))
+				h.AssertPathDoesNotExist(t, filepath.Join(layerDir, "metadata.buildpack", "cache-false.sha"))
+			})
+
+			it("restores escaped buildpack layer metadata", func() {
+				err := layerAnalyzer.Analyze(buildpacks, skipLayers, layersMetadata, cacheMetadata)
+				h.AssertNil(t, err)
+
+				path := filepath.Join(layerDir, "escaped_buildpack_id", "escaped-bp-layer.toml")
+				got := h.MustReadFile(t, path)
+				want := "[metadata]\n  escaped-bp-layer-key = \"escaped-bp-layer-value\""
+
+				h.AssertStringContains(t, string(got), want)
+			})
+
+			it("restores each store metadata", func() {
```

Comment:
Is this a new assertion?

Reply:
Yeah, sort of. It's really a continuation of existing tests, but there wasn't one for exactly this situation. It might've been covered by other tests, but I figured the explicit test would be an improvement.